src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationCodeFixProvider.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.RemoveUnnecessaryDiscardDesignation
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp,
+        Name = nameof(RemoveUnnecessaryDiscardDesignationCodeFixProvider)), Shared]
+    internal partial class RemoveUnnecessaryDiscardDesignationCodeFixProvider : SyntaxEditorBasedCodeFixProvider
+    {
+        [ImportingConstructor]
+        [SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
+        public RemoveUnnecessaryDiscardDesignationCodeFixProvider()
+        {
+        }
+
+        public override ImmutableArray<string> FixableDiagnosticIds
+            => ImmutableArray.Create(IDEDiagnosticIds.RemoveUnnecessaryDiscardDesignationDiagnosticId);
+
+        internal sealed override CodeFixCategory CodeFixCategory => CodeFixCategory.CodeStyle;
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var diagnostic = context.Diagnostics.First();
+            context.RegisterCodeFix(
+                new MyCodeAction(c => FixAsync(context.Document, diagnostic, c)),
+                diagnostic);
+
+            return Task.CompletedTask;
+        }
+
+        protected override Task FixAllAsync(
+            Document document, ImmutableArray<Diagnostic> diagnostics,
+            SyntaxEditor editor, CancellationToken cancellationToken)
+        {
+            var generator = editor.Generator;
+
+            foreach (var diagnostic in diagnostics)
+            {
+                var discard = diagnostic.Location.FindNode(getInnermostNodeForTie: true, cancellationToken);
+                switch (discard.Parent)
+                {
+                    case DeclarationPatternSyntax declarationPattern:
+                        if (declarationPattern.Parent is IsPatternExpressionSyntax isPattern)
+                        {
+                            editor.ReplaceNode(
+                                isPattern,
+                                (current, _) =>
+                                {
+                                    var currentIsPattern = (IsPatternExpressionSyntax)isPattern;
```

Comment:
Why is this necessary?

Reply:
when you have `x is T y` that's an `IsPatternSyntax`.  And an IsPattern is *not* a BinaryExpression.  However, are converting this to `x is T`.  In that case, that's a BinaryExpression.  So we have to specially break this apart and reform into a BinaryExpr. 

Should `x is T` be a BinaryExpr?  Probably not.  But we shipped that way 10 years ago, and so here we are :)

Reply:
I meant why do we need to cast `isPattern` when it looks like it already is the right type, and it doesn't look like the value of it changes, and there is other code in this lambda that uses it instead of `currentIsPattern`.

Not that the explanation you gave wasn't interesting :)

Reply:
yeah.  that wa a bug :)  but have fixed in the code that got merged.  sorry i didn't realize it!